modules/entity_version_history/tests/src/Kernel/HistoryTabTest.php

```diff
@@ -0,0 +1,154 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\Tests\entity_version_history\Kernel;
+
+use Drupal\Core\Url;
+use Drupal\KernelTests\KernelTestBase;
+use Drupal\Tests\user\Traits\UserCreationTrait;
+use Drupal\user\Entity\User;
+use Symfony\Component\Routing\Exception\RouteNotFoundException;
+use Symfony\Component\Routing\Route;
+
+/**
+ * Tests the entity version history tab.
+ *
+ * @group entity_version_history
+ */
+class HistoryTabTest extends KernelTestBase {
+
+  use UserCreationTrait;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected static $modules = [
+    'user',
+    'node',
+    'field',
+    'text',
+    'system',
+    'entity_version',
+    'entity_version_history',
+    'entity_version_history_test',
+  ];
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp() {
+    parent::setUp();
+
+    $this->installEntitySchema('node');
+    $this->installEntitySchema('entity_version_history_settings');
+
+    $this->installConfig([
+      'user',
+      'node',
+      'system',
+      'field',
+      'entity_version',
+      'entity_version_history',
+      'entity_version_history_test',
+    ]);
+
+    // Create a history tab setting for the corresponding entity type
+    // and bundle.
+    $history_storage = $this->container->get('entity_type.manager')
+      ->getStorage('entity_version_history_settings');
+    $history_storage->create([
+      'target_entity_type_id' => 'node',
+      'target_bundle' => 'first_bundle',
+      'target_field' => 'field_entity_version',
+    ])->save();
+
+    $this->container->get('entity_type.manager')->clearCachedDefinitions();
+  }
+
+  /**
+   * Tests that the History tab is correctly applied.
+   */
+  public function testHistoryTabApplied(): void {
+    $route_provider = $this->container->get('router.route_provider');
+
+    /** @var \Drupal\Core\Entity\EntityTypeInterface $definition */
+    foreach ($this->container->get('entity_type.manager')->getDefinitions() as $definition) {
+      if ($definition->id() === 'node') {
+        $this->assertTrue($definition->hasLinkTemplate('drupal:entity-version-history'));
+        $this->assertInstanceOf(Route::class, $route_provider->getRouteByName('entity.' . $definition->id() . '.history'));
+        continue;
+      }
+
+      $this->assertFalse($definition->hasLinkTemplate('drupal:entity-version-history'));
+      try {
+        $route_provider->getRouteByName('entity.' . $definition->id() . '.history');
+      }
+      catch (\Exception $exception) {
+        $this->assertInstanceOf(RouteNotFoundException::class, $exception);
+      }
+    }
+  }
+
+  /**
+   * Tests access to the history tab.
+   */
+  public function testHistoryTabAccess(): void {
+    $this->installEntitySchema('user');
+    // Define the anonymous user first.
+    User::create([
```

Comment:
Why aren't you using the UserCreationTrait you added for making users with various permissions? You wont even need anonymous user but one without a permission and one with.

Reply:
That's a good idea to have 2 users instead. However User id with 0 has to exist in order to avoid "ContextException: The 'entity:user' context is required and not present" error. Core does the same in their Kernel tests.